pytensor/compile/builders.py

```diff
@@ -469,327 +416,254 @@ def __str__(self):
-                self._lop_op_stypes_l = [None] * inp_len
-                self._lop_op.kwargs["on_unused_input"] = "ignore"
-                return
-
-        output_grads = [out_t() for out_t in self.output_types]
+        ]
         fn_grad = partial(
             grad,
             cost=None,
             disconnected_inputs="ignore",
             return_disconnected="Disconnected",
             null_gradients="return",
-            known_grads=OrderedDict(zip(local_outputs, output_grads)),
+            known_grads=OrderedDict(zip(inner_outputs, output_grads)),
         )
 
-        assert self._lop_type in ("lop", "grad"), self.LOP_TYPE_ERR_MSG % self._lop_type
-        if self._lop_type == "lop":
-            callable_args = (local_inputs, local_outputs, output_grads)
-        elif self._lop_type == "grad":
-            callable_args = (local_inputs, output_grads)
+        if self._lop_op_interface:
+            callable_args = (inner_inputs, inner_outputs, output_grads)
+        else:
+            callable_args = (inner_inputs, output_grads)
 
         # we need to convert _lop_op into an OfG instance
-        if lop_op == "default":
-            gdefaults_l = fn_grad(wrt=local_inputs)
-            all_grads_l, all_grads_ov_l = zip(
-                *[
-                    OpFromGraph._filter_grad_var(grad, inp)
-                    for grad, inp in zip(gdefaults_l, local_inputs)
-                ]
-            )
-            all_grads_l = list(all_grads_l)
-            all_grads_ov_l = list(all_grads_ov_l)
-        elif isinstance(lop_op, Variable):
-            if isinstance(lop_op.type, DisconnectedType | NullType):
-                all_grads_l = [inp.zeros_like() for inp in local_inputs]
-                all_grads_ov_l = [lop_op.type() for _ in range(inp_len)]
-            else:
-                raise ValueError(self.STYPE_ERR_MSG % lop_op.type)
-        elif isinstance(lop_op, list):
-            goverrides_l = lop_op
-            if len(goverrides_l) != inp_len:
+        if lop_op_constructor is None:
+            input_grads = fn_grad(wrt=inner_inputs)
+        elif isinstance(lop_op_constructor, list):
+            custom_input_grads = lop_op_constructor
+            if len(custom_input_grads) != nin:
                 raise ValueError(
-                    f"Need to override {int(inp_len)} gradients, got {len(goverrides_l)}",
-                    goverrides_l,
+                    f"Need to override {nin} gradients, got {len(custom_input_grads)}",
+                    custom_input_grads,
                 )
             # compute non-overriding downsteam grads from upstreams grads
             # it's normal some input may be disconnected, thus the 'ignore'
-            wrt_l = [
-                lin for lin, gov in zip(local_inputs, goverrides_l) if gov == "default"
+            wrt = [
+                lin
+                for lin, gov in zip(inner_inputs, custom_input_grads)
+                if gov == "default"
```

Comment:
Use `None` in these lists to indicate no gradient override instead of string `"default"`? Matches the changes to the outer-most API.

Could accept both for now and give a FutureWarning for `default` if you're worried about legacy code

Reply:
I think None was avoided because it's confusing with NullType/DisconnectedType/No gradient. This is not a concern at the global level, so I changed it only there

Reply:
Can you pass those in the list of lop_overrides?

Reply:
> Can you pass those in the list of lop_overrides?

Yup, it's the only variable you're allowed to pass. Otherwise must be a callable or a "default"

Reply:
But None makes sense in the context of overrides, where you are NOT overriding. The API would look like this, to force an output gradient to be disconnected, the other to be custom, and a final one to use the default mahcinery:

```python
lop_overrides = [disconnected_type(), custom_lop_callable, None]
```

What do you think? This is very niche so I am okay with a less user-friendly API

Reply:
Agreed let's add that to this PR TODO list then

Reply:
Maybe I would simplify this even further and only allow None or callable. The user can return `disconnected_type` from the callable easily enough?

Reply:
I already removed the option to say `lop_overrides=disconnected_type()` which is an implicit `lop_overrides=[disconnected_type() for i in range(len(inputs))]`

Reply:
NVM, no need to touch it if it ain't broken I guess. I'll just get rid of the "default" string

Reply:
Done